src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,748 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+
+/**
+ * Represents a string.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    protected $string = '';
+    protected $ignoreCase = false;
+
+    abstract public function __construct(string $string = '');
```

Comment:
Seems like this abstract class would be more flexible if the constructor was defined:
```php
function __construct(string $string = '') {
	$this->string = $string;
}
```

I don't think the abstract should tell the parent how to construct itself, the string can come from anywhere. Instead the parent should pass in the required constructor-args via `parent::__construct()`

Reply:
I made this to enforce a consistent constructor signature.

Reply:
This could still be enforced internally, but don't see the point of enforcing it for user-classes.

Reply:
> This could still be enforced internally

WDYM?

Reply:
I just mean that you could keep the signatures as is for the parent classes in this lib (weak enforcement). But still allow user-classes to choose their own signature for their classes.

The only thing this really enforces is having `$string` as first argument, and all other constructor-args to have default values.

Also this prevents having private constructors w/ named constructors in users classes (would still be a problem in 7.1).

Reply:
I want to be able to do `new static($string)` in the abstract class, that's why.

Reply:
Never thought of having abstract on a constructor of an abstract class... very smart.. first time I see it and TBH it kinda feels the same as having an interfaces with a constructor.

> I want to be able to do new static($string) in the abstract class, that's why.

I think this reminds me of `Collections` from doctrine.. According to the doc this might be the reason they added `createFrom` function... 

https://github.com/doctrine/collections/blob/master/lib/Doctrine/Common/Collections/ArrayCollection.php#L95

ofc since there is no interface here current solution might be fine. Your call :)

Reply:
> I want to be able to do `new static($string)` in the abstract class, that's why.

Ah, I missed that. However it's only used in [one place](https://github.com/symfony/symfony/pull/33553/files#diff-2ddd8d035114516a9fa39ede6812d652R625) and could be replaced with a clone.
Also the constructor abstract doesn't enforce that `$this->string` is actually being set nor with what value.

That being said, I guess it's a good enough reason so whatever you want to do.

Reply:
It's used once today but could be more in the future, that's an extensibility hook :)

Reply:
Regarding the discussion about the `ignoreCase` situation below, wouldn't it be nice to have the ability to pass an AbstractString object itself too? This way the whole situation about immutability and being able to change an internal flag of such immutable object could be handled simply by saying we have a first argument of type string and AbstractString and a second argument telling the object to handle everything ignoring the case or not.